Announce automatic checking of cfgs at compile-time (#1313)
rust-lang/blog.rust-lang.org: posts/2024-05-06-check-cfg.md

author: Urgau
team: The Cargo Team <https://www.rust-lang.org/governance/teams/dev-tools#cargo>
---

The Cargo and Compiler team are delighted to announce that starting with Rust 1.80 (or nightly-2024-05-05) every _reachable_ `#[cfg]` will be **automatically checked** that they match the **expected config names and values**.

This can help with verifying that the crate is correctly handling conditional compilation for different target platforms or features. It ensures that the cfg settings are consistent between what is intended and what is used, helping to catch potential bugs or errors early in the development process.

This addresses a common pitfall for new and advanced users.

This is another step to our commitment to provide user-focused tooling and we are eager and excited to finally see it fixed, after more than two years since the original [RFC 3013](https://github.com/rust-lang/rfcs/pull/3013)[^1].

[^1]: The stabilized implementation and RFC 3013 diverge significantly, in particular there is only one form for `--check-cfg`: `cfg()` (instead of `values()` and `names()` being incomplete and subtlety incompatible with each other).

## A look at the feature 

> Every time a Cargo feature is declared that feature is transformed into a config that is passed to `rustc` (the Rust compiler) so it can verify with its [well known cfgs](https://doc.rust-lang.org/nightly/rustc/check-cfg.html#well-known-names-and-values) if any of the `#[cfg]`, `#![cfg_attr]` and `cfg!` have unexpected configs and report a warning with the `unexpected_cfgs` lint.

Reviewer: Do you mean

```suggestion
Every time a Cargo feature is declared that feature is transformed into a config that is passed to `rustc` (the Rust compiler) so it can verify with it along with [well known cfgs](https://doc.rust-lang.org/nightly/rustc/check-cfg.html#well-known-names-and-values) if any of the `#[cfg]`, `#![cfg_attr]` and `cfg!` have unexpected configs and report a warning with the `unexpected_cfgs` lint.
```

[^2]: `cargo::rustc-check-cfg` will start working in Rust 1.80 (or nightly-2024-05-05). From Rust 1.77 to Rust 1.79 *(inclusive)* it is silently ignored. In Rust 1.76 and below a warning is emitted when used without the unstable Cargo flag `-Zcheck-cfg`.

The syntax to use is described in the [rustc book](https://doc.rust-lang.org/nightly/rustc/) section [checking configuration](https://doc.rust-lang.org/nightly/rustc/check-cfg.html), but in a nutshell the basic syntax of `--check-cfg` is:

```
cfg(name, values("value1", "value2", ..., "valueN"))
```

Note that every custom cfgs must always be expected, regardless if the cfg is active or not!

### `build.rs` example

`build.rs`:
```rust
fn main() {
    println!("cargo::rustc-check-cfg=cfg(has_foo)");
    //        ^^^^^^^^^^^^^^^^^^^^^^ new with Cargo 1.80
    if has_foo() {
        println!("cargo::rustc-cfg=has_foo");
    }
}
```

> > Each `cargo::rustc-cfg` must always have a accompanying **unconditional** `cargo::rustc-check-cfg` directive otherwise you risk having warnings like this: `unexpected cfg condition name: has_foo`.

Reviewer: ```suggestion
> Each `cargo::rustc-cfg` should have an accompanying **unconditional** `cargo::rustc-check-cfg` directive to avoid warnings like this: `unexpected cfg condition name: has_foo`.
```
- grammar
- The "must" language makes this sound like its generating errors. Yes, it later says warning but setting expectations early in the statement would help

### Equivalence table

| `cargo::rustc-cfg`      | `cargo::rustc-check-cfg`                       |
|-------------------------|------------------------------------------------|
| `foo`                   | `cfg(foo)` or `cfg(foo, values(none()))`       |
| `foo=""`                | `cfg(foo, values(""))`                         |
| `foo="bar"`             | `cfg(foo, values("bar"))`                      |
| `foo="1"` and `foo="2"` | `cfg(foo, values("1", "2"))`                   |
| `foo="1"` and `bar="2"` | `cfg(foo, values("1"))` and `cfg(bar, values("2"))` |
| `foo` and `foo="bar"`   | `cfg(foo, values(none(), "bar"))`              |

More details can be found in the [`rustc` book](https://doc.rust-lang.org/nightly/rustc/check-cfg.html).

## Frequently asked questions

### Can it be disabled?

> The feature is **always on** and _cannot_ be disabled from Cargo, but like any other lints it can be controlled: `#![warn(unexpected_cfgs)]`.

Reviewer: ```suggestion
For Cargo users, the feature is **always on** and _cannot_ be disabled, but like any other lints it can be controlled: `#![warn(unexpected_cfgs)]`.
```

### Does the lint affect dependencies?

> No, the lint only affects local packages; only those will report the lint.

Reviewer: ```suggestion
No, like most lints, `unexpected_cfgs` will only be reported for local packages thanks to [cap-lints](https://doc.rust-lang.org/rustc/lints/levels.html#capping-lints).
```
